### ISSUE TITLE
Move macwire away from adamw/softwaremill to its own organization

### ISSUE THREAD

**Issue by backuitist** [adamw/macwire#80]
Packages and build organization should be renamed into `org.macwire`. The project should be moved to a `macwire` Github organization. 
I guess that, as it will be a new artifact, it'll have version 1.0.0.

**Comment by backuitist** [comment]
I've created https://github.com/mac-wire as `macwire` was already taken.

**Comment by adamw** [comment]
What's wrong with the current organization or account? It would also require getting a domain. I'm not sure it's worth confusing the users. There's a lot of popular projects on normal user accounts, no?

**Comment by backuitist** [comment]
You don't really need a domain. You can use a github subdomain, very much like di-in-scala.
About the current organization (and user account for that matter), it's clearly bringing attention to softwaremill (and adamw ;)), which is fine as long as you're the only committer, or as long as the only committers are softwaremill-ers ;)
Most well-known open source projects aren't affiliated to a company because that would be unfair to the other companies/individuals that are also contributing (pants build is a good example). Plus you never know, tomorrow you might work in a different company or simply decide to rebrand your company (like typesafe).
Once macwire is on its own, we can add a credit section.

**Comment by adamw** [comment]
Well you need a domain if you want to deploy to maven central (as far as I know).

Anyway, I don't think that it was an issue so far, so I'm going to leave things as-is for now. If there are more users interested in such a move - this is a good place to voice your opinion :). I definitely don't want to scare any potential contributors away, but I hope that the account name doesn't matter that much really.

Moving to a separate organisation would make sense if there were several sub-projects (which I don't think is happening at least yet) or if this was a really huge "framework" on top of which you build applications. I still view macwire as a rather small helper library to make a programmer's life easier :)

**Comment by backuitist** [comment]
> I definitely don't want to scare any potential contributors away, but I hope that the account name doesn't matter that much really.

Sure, it's not the account name that bothers me the most, it's the organization. Here is what you see when you land on adamw/macwire: "Lightweight and Nonintrusive Scala Dependency Injection Library
http://www.softwaremill.com". So when you start contributing significantly, you wonder, do I work for softwaremill?

> If there are more users interested in such a move

This is mostly about contributors, not users. As a user I don't really care if that's 'com.jujubes.macwire' or 'org.mydearcharity.macwire', as long as I can get the code with a permissive licence.

> Well you need a domain if you want to deploy to maven central (as far as I know).

See http://central.sonatype.org/pages/choosing-your-coordinates.html
"Other examples are [...] github.com/yourusername -> com.github.yourusername"

Try a search: https://search.maven.org/#search|ga|1|com.github
"Displaying 1 to 20 of 6639" - a fairly common practice.

**Comment by backuitist** [comment]
> Moving to a separate organisation would make sense if there were several sub-projects

A good start would be a documentation website. Like I said in #60 I don't really like the state of affairs - hence my scarce contributions to the README.

> I still view macwire as a rather small helper library

Size does not matter ;)

**Comment by backuitist** [comment]
I guess you've already been there, but that's another good example: https://github.com/scaldi
A small scala DI library with its own organization.

**Comment by adamw** [comment]
Hmm why would you be working for softwaremill? It's the company I'm working at and I use that org for most open-source that I write. The same I don't have trouble contributing to `com.google`, I don't see why this would be a problem here :) I'm sure there's a lot of projects using the github domain, but this would be just confusing to users without a real benefit

As for docs, I think it's best if they live alongside the code, why would we want a separate repository :)

**Comment by backuitist** [comment]
> Hmm why would you be working for softwaremill?

Exactly, why? Softwaremill is the company you co-founded and you're obviously using open-source to promote it. Why don't you admit it? There's nothing wrong with that.
Like you I have no problem making small contributions to open-source projects that are advertising a lucrative company, but I feel differently when I'm rewriting the whole thing.

> As for docs, I think it's best if they live alongside the code, why would we want a separate repository :)

You mean as a readme? I don't think that format is adequate, plus it's fairly easy to create a jenkyll documentation website with a decent navigation.

> I definitely don't want to scare any potential contributors away

I've started this whole discussion because that's bothering me, so don't take my contributions for granted. I've actually done quite some work over the last few days that is on hold until this gets fixed.

**Comment by backuitist** [comment]
> this would be just confusing to users without a real benefit

I don't think that at this stage (how many projects are using macwire in production?) it'll confuse people that much. We're talking about renaming some imports and a dependency in your build configuration.

**Comment by adamw** [comment]
> Exactly, why? Softwaremill is the company you co-founded and you're obviously using open-source to promote it. Why don't you admit it? There's nothing wrong with that.
Like you I have no problem making small contributions to open-source projects that are advertising a lucrative company, but I feel differently when I'm rewriting the whole thing.

Sure, it is promoting Softwaremill in some way, though I don't think it's really significant. And I don't think anybody views contributing to a library created by some company as working for that company. Also I think there's quite some code left from the original ;)

> You mean as a readme? I don't think that format is adequate, plus it's fairly easy to create a jekyll documentation website with a decent navigation.

Not necessarily, it can be `.md` files in a `docs` subdirectory or anything else

> I don't think that at this stage (how many projects are using macwire in production?) it'll confuse people that much. We're talking about renaming some imports and a dependency in your build configuration.

I don't have any numbers, but I think quite a few. As for now I'm not going to do the rename - sorry.

> I've started this whole discussion because that's bothering me, so don't take my contributions for granted. I've actually done quite some work over the last few days that is on hold until this gets fixed.

Look, I'll be happy to merge more changes (as long as they make sense ;) ) but I'm also not going to do everything just to get them in. I really wouldn't like to see you go, and I'm not sure why this is so important suddenly, it's a simple open-source project, let's make sure it's as good as it can be and keep coding :)

**Comment by backuitist** [comment]
> Sure, it is promoting Softwaremill in some way, though I don't think it's really significant.

If that's not significant, why not doing it? I'd be happy to help you move the project, and create a good documentation.

> I'm also not going to do everything just to get them in

Is that so much to ask? So you'd rather invest your time on coding and maintaining than simply transitioning the project to a more open and independent space where you can benefit, for free, from my contributions?

Take akka for example, it used to be se.scalablesolutions.akka. See https://github.com/akka/akka/commit/680ee7ddf105fa5c7057f30a899c4cad3a36a665

**Comment by ktoso** [comment]
I'm not sure using Akka as an argument fits well into the discussion here.
Technically speaking akka is strictly published under the `com.typesafe.akka` groupId.
Packages inside the artifacts indeed `akka.*`, however this is for convenience / style.

Personally, I've been happily contributing to `com.twitter.*` or `google` or `foursquare` repositories with an "they started it, so it lives there" philosophy, and the situation here seems to be pretty similar.

Just my 2 cents, don't want to dive too deep into the discussion here – it's something the maintainer is ought to decide himself.

**Comment by backuitist** [comment]
> I'm not sure using Akka as an argument fits well into the discussion here.

The argument was that renaming the packages would be confusing to the users. I'm bringing Akka as an example of how a well-established project with a far larger user base was able to do such a change.

> Personally, I've been happily contributing to com.twitter.* or com.google or com.foursquare or "someone's random repo" repositories

Again the question is about how much you contribute. I've also been contributing marginally to `com.google`, `sun`, ...

> the situation here seems to be pretty similar.

When you're responsible for [> 50% of the current code of the whole project](https://github.com/adamw/macwire/blame/master/macros/src/main/scala/com/softwaremill/macwire/dependencyLookup/EligibleValuesFinder.scala) it seems to me that you might be entitled to question how the project is used, and if some advertisement is derived from it, you should rightfully have a word to say.

> it's something the maintainer is ought to decide himself.

Of course, I'm not forcing anyone... and nobody is forcing me to contribute.

**Comment by adamw** [comment]
> If that's not significant, why not doing it? I'd be happy to help you move the project, and create a good documentation.

That was just a side note, not really important.

> Is that so much to ask? So you'd rather invest your time on coding and maintaining than simply transitioning the project to a more open and independent space where you can benefit, for free, from my contributions?

I suspect that I would be maintaining the project anyway ;) I think the project is quite open, the code is there, all PRs are reviewed etc., so let's not create artificial problems.

Look, if you'd like to somehow underline that you are not from Softwaremill, we can create a "contributors" section with a link of your choosing or something like that.

**Comment by backuitist** [comment]
> I suspect that I would be maintaining the project anyway ;)

If you leave things as they are it's very likely. Otherwise I wouldn't mind maintaining it - as a matter of fact I'm already maintaining my own private version, and it's interesting to see that you're more or less (re)producing the same code as I did a week ago.

>  we can create a "contributors" section

Yep, that's what I suggested, once it's moved to its organization, only that it would be called "credit".
See https://pantsbuild.github.io/credits.html for inspiration ;)

**Comment by adamw** [comment]
> If you leave things as they are it's very likely. Otherwise I wouldn't mind maintaining it - as a matter of fact I'm already maintaining my own private version, 

You make it sound as if things were in a bad state. Sorry, but I really don't believe that the package name/organization can be such a problem. EOT on my side.

> and it's interesting to see that you're more or less (re)producing the same code as I did a week ago.

which only validates the approach ;)

> Yep, that's what I suggested, once it's moved to its organization, only that it would be called "credit".
See https://pantsbuild.github.io/credits.html for inspiration ;)

Well I guess that's a 5-minute job, go ahead :)

**Comment by backuitist** [comment]
> go ahead :)

Two things:
* i'm on holiday for 2 weeks :)
* if you don't agree, there won't be 2 versions of macwire on github, that would definitely create some confusion, plus that would sound like a fight, whereas I just want to work with you on a fair basis (at least fair in my opinion)

**Comment by adamw** [comment]
> i'm on holiday for 2 weeks :)

Have a good time :)

> if you don't agree, there won't be 2 versions of macwire on github, that would definitely create some confusion, plus that would sound like a fight, whereas I just want to work with you on a fair basis (at least fair in my opinion)

I'm not sure I understand this ;) Anyway, I guess I said everything that I wanted to here

**Comment by backuitist** [comment]
> I'm not sure I understand this ;)

It's simple, if you agree with the organization move, then the following (amongst other things) should be done:
* move the project to `github.com/macwire/macwire`
* have a link on `github.com/adamw/macwire` (at that point this repo will be empty), redirecting to `macwire/macwire`

When you said 'go ahead' I understood that you told me to fork your project on github - which I won't as I said previously.

**Comment by adamw** [comment]
No I meant adding the credits section.
I think I wrote a couple of time that I'm not going to change the organisation as of now.

**Comment by backuitist** [comment]
I just created the macwire organization (that was blocked by a phantom user BTW). Now that one year has passed you might want to reconsider this?

**Comment by tkroman** [comment]
@backuitist wow, looks like this is really important for you. Besides the enterprise-backed `com.twitter|google|whatever`, there are also lots of personal repos like `non/spire`, `milessabin/shapeless` etc, and both maintainers and contributors are fine with that. So, if no one raised this issue in the past year and the repo owner expressed his unwillingness pretty clearly, is there really a point in reiterating on this?
Usual disclaimer on 'my 2 cents' here.

**Comment by backuitist** [comment]
I can reiterate my points for you: First, I don't feel comfortable with the advertisement done on the project (first and foremost link points to softwaremill), which IMO is not comparable with the other projects that you mentioned.
Second, it's a matter of involvement. It's funny that you consider the contributors as being fine with this although, to this date, if you remove softwaremill employees, you're basically left with me and some extremely marginal contributors ;)

The scala ecosystem is fragmented enough (look at json libraries) that it seemed wiser not to fork this project. But I guess I'll follow your advice and fork so I can release my work on better terms.

**Comment by adamw** [comment]
The link will be actually changed soon to point to our open-source page, which is in the works, though I don't think it really matters to that discussion. Is the link really such a problem? I find that hard to believe :)

As for involvement, if all major contributors are indeed SoftwareMill employees and you (haven't checked that), how is that an argument against using the `com.softwaremill` package?

It's open-source so you are free to fork of course, I'll be curious to check out what kind of improvements you have made, and maybe back-port some (retaining all attributions of course). And so there's no doubt: I'd be even more happy to get some PRs to this project.
